projects/graph/graph.py

```diff
@@ -11,45 +11,145 @@ def add_vertex(self, vertex):
+                    s.push(neighbor)
+
+
+
+
     def dft_recursive(self, starting_vertex):
         """
         Print each vertex in depth-first order
         beginning from starting_vertex.
         This should be done using recursion.
         """
         pass  # TODO
+    
+    
+    
+    
     def bfs(self, starting_vertex, destination_vertex):
         """
         Return a list containing the shortest path from
         starting_vertex to destination_vertex in
         breath-first order.
         """
-        pass  # TODO
+        # Create an empty set to store visited nodes
+        visited = set()
+        # Create an empty Queue and enqueue A PATH TO the starting vertex
+        q = Queue()
+        q.enqueue( [starting_vertex] )
+        # While the queue is not empty...
+        while q.size() > 0:
+            # Dequeue the first PATH
+            path = q.dequeue()
+            # GRAB THE VERTEX FROM THE END OF THE PATH
+            v = path[-1]
+            # IF VERTEX == TARGET, RETURN PATH
+            if v == destination_vertex:
+                return path
+            # If that vertex has not been visited...
+            if v not in visited:
+                # Mark it as visited
+                visited.add(v)
+                # Then add A PATH TO all of its neighbors to the back of the queue
+                for neighbor in self.vertices[v]:
+                    # Copy the path
+                    path_copy = list(path)
+                    # Append neighbor to the back of the copy
+                    path_copy.append(neighbor)
+                    # Enqueue copy
+                    q.enqueue(path_copy)
+
+
+
     def dfs(self, starting_vertex, destination_vertex):
         """
         Return a list containing a path from
         starting_vertex to destination_vertex in
         depth-first order.
         """
-        pass  # TODO
+        # Create an empty set to store visited nodes
+        visited = set()
+        # Create an empty Stack and push A PATH TO the starting vertex
+        s = Stack()
+        s.push( [starting_vertex] )
+        # While the stack is not empty...
+        while s.size() > 0:
+            # Pop the first PATH
+            path = s.pop()
+            # GRAB THE VERTEX FROM THE END OF THE PATH
+            v = path[-1]
+            # IF VERTEX == TARGET, RETURN PATH
+            if v == destination_vertex:
+                return path
+            # If that vertex has not been visited...
+            if v not in visited:
+                # Mark it as visited
+                visited.add(v)
+                # Then add A PATH TO all of its neighbors to the top of the stack
+                for neighbor in self.vertices[v]:
+                    # Copy the path
+                    path_copy = list(path)
+                    # Append neighbor to the back of the copy
+                    path_copy.append(neighbor)
+                    # Enqueue copy
+                    s.push(path_copy)
 
```

Comment:
Nice work with the BF/DF searches and traversals.  Isn't the code so similar for the two of them?

Not for the sake of this project... but I bet we could write a function that handles both BF and DF search cases based on what parameter is passed in.